rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+
+#include "logging.hpp"
+#include "rosbag2_compression/zstd_compressor.hpp"
+
+namespace rosbag2_compression
+{
+
+std::string ZstdCompressor::compress_uri(const std::string & uri)
+{
+  auto start = std::chrono::high_resolution_clock::now();
+  auto compressed_uri = uri + "." + get_compression_identifier();
+  std::ifstream infile(uri);
+  if (infile.is_open()) {
+    // Get size and allocate
+    infile.seekg(0, std::ios::end);
+    size_t decompressed_buffer_length = infile.tellg();
+    char * decompressed_buffer = new char[decompressed_buffer_length];
```

Comment:
```c++
auto decompressed_buffer = std::make_unique<char>(decompressed_buffer_length);
```

Reply:
what about using std::string instead?

Reply:
`decompressed_buffer` is a memory block. It is possible to use `std::string`, however the purpose of `std::string` is for text.

For science, it looks like the solution using `std::string` would be:
```c++
std::string decompressed_buffer{decompressed_buffer_length, '\0'};

function_that_accepts_buffer(decompressed_buffer.data());
```

Reply:
On my radar

Reply:
instead of `char` I would recommend `uint8_t` as its size is clearly defined on all platforms.
Also, this allocation needs a check whether it's successfully allocated the memory to avoid segfaults in case it's not.
https://stackoverflow.com/questions/31090616/printf-adds-extra-ffffff-to-hex-print-from-a-char-array